### ISSUE TITLE
Unable to use future_syntax with dune 1.10.0

### ISSUE THREAD

**Issue by Dean177** [ocaml/dune#2262]
I am attempting to use the new `ocaml-syntax-shims` with other preprocessors on dune 1.10.0 

but get the following error:
```
 (pps ppx_inline_test -- -pp %{bin:ocaml-syntax-shims})
                                         ^^^^^^^^^^^^^^^^^^^^^^^
Error: %{bin:..} isn't allowed in this position
```

Reproducing example:

```dune
(library
  (name future_example)  
  (inline_tests)
  (preprocess (pps ppx_inline_test -- -pp %{bin:ocaml-syntax-shims})))
```

```ocaml
let (let+) l f = 
  List.map f l

let add_one l = 
  let+ elem = l in
  elem + 1

let%test _ =
  add_one [1;2;3] = [2;3;4]

```
See also #1934

**Comment by aalekseyev** [comment]
Apparently we don't expand %{bin} here, probably for no particular reason.

As a workaround you could try something like this:

```
(library
 (name future_example)
 (preprocess (pps ppx_inline_test -- -pp ./ocaml-syntax-shims))
 (preprocessor_deps ./ocaml-syntax-shims)
 )
(rule (copy %{bin:ocaml-syntax-shims} ./ocaml-syntax-shims))
```

@mlasson, I'm told you know something about ppx flag expansion.
Do you think it would make sense to support more variables in ppx flag expansion?

**Comment by Dean177** [comment]
Thanks @aalekseyev, that works perfectly

**Comment by mlasson** [comment]
@aalekseyev : Yes it would totally makes sense, but I think some thinking should be put into that.

Currently, most macro are allowed only in actions (except for %{env} which feels like a hack).
Moreover, these expansions in flags are "static": they are triggered even if you don't build a thing and %{bin} could fail (because the binary is not in the installed world). 
 
If we wanted to support more macros in preprocessors flags we would need to change the logic in order to delay the expansion at the build stage. Moreover, for consistency, we would probably want to do that for other kinds of flags. So this does not seem trivial to me (but I'm far from being an expert).

**Comment by emillon** [comment]
I'm not sure about the timeline here, but this has been fixed in #2076 (dune 1.10). Using `(lang dune 1.10)` enables this.
